### ISSUE TITLE
[mobilebert] Tensor had NaN values Error when using quantization-aware training 

### ISSUE THREAD

**Issue by zaidalyafeai** [google-research/google-research#384]
Hey, I am trying to retrain mobilebert. I was able to run the pretraing scrips with no problems. However, when fine-tuning on squad using the flag `--use_quantized_training=true`. I get the following error 

```
ERROR:tensorflow:Error recorded from training_loop: From /job:worker/replica:0/task:0:
Gradient for bert/encoder/layer_9/attention/output/dense/bias:0 is NaN : Tensor had NaN values
```
Here is the script I use for fine-tuning 

```
python3 run_squad.py \
	--bert_config_file=config/uncased_L-24_H-128_B-512_A-4_F-4_OPT_QAT.json \
	--data_dir=${DATA_DIR} \
	--do_train \
	--doc_stride=128 \
	--init_checkpoint={INIT_CKPT}/model.ckpt-1000.index \
	--learning_rate=4e-05 \
	--predict_file=dev.json \
	--do_lower_case \
  	--do_predict \
	--max_answer_length=30 \
	--max_query_length=64 \
	--max_seq_length=128\
	--n_best_size=20 \
	--num_train_epochs=1 \
	--output_dir=${OUTPUT_DIR} \
	--train_batch_size=32 \
	--train_file=train.json \
	--use_tpu \
	--tpu_name=${TPU_NAME} \
	--vocab_file=../bert/en-vocab.txt \
	--warmup_proportion=0.1 \
	--verbose_logging=True \
	--use_quantized_training=true
```

**Comment by tromedlov22** [comment]
Replace 
--init_checkpoint={INIT_CKPT}/model.ckpt-1000.index \
by
--init_checkpoint={INIT_CKPT}/model.ckpt-1000 \

**Comment by UmarSpa** [comment]
Hi @zaidalyafeai @tromedlov22 
Sorry to bother you. Since you are already working on this, I was wondering if you can help me with my following queries:

- For the data preparation (python create_pretraining_data.py) I am using the [code](https://github.com/google-research/bert/blob/master/create_pretraining_data.py) in BERT repo. I have wiki dataset (13G) and I have sharded it and created 130 small txt files. But when I run the code, it takes almost 1 day for each sharded file. Is there anything I am doing wrong ? I am using max_seq_length=512, as this is needed to replicate the results.

- For mobilebert, do you generate two separate sets of tf.records using the above code. One with max_seq_length=128 and other with max_seq_length=512 ?

**Comment by zaidalyafeai** [comment]
Hi @UmarSpa 

1. The script doesn't deal with large text files. For my case, I had to split the file into multiple files each with size around ~ 500MB `split -l 100000 files` then you can use a for loop like this one 

```
for file in glob.glob('../train_files/**.txt'):
  !python3 create_pretraining_data.py \
    --input_file={file}.txt \
    --output_file=path/to/output/{file}.tfrecord \
    --vocab_file=vocab.txt \
    --max_seq_length=512 \
    --max_predictions_per_seq=20 \
    --masked_lm_prob=0.15 \
    --random_seed=12345 \
    --dupe_factor=5
```

2. To my understanding mobilebert didn't use two stages of training since `first_num_train_steps=0`. The two stages training are used to speed up the training, in either cases you don't need to create two sets of pretraining data because the code automatically truncate the sequences to `128`.

**Comment by UmarSpa** [comment]
@zaidalyafeai thanks for the reply.

Data split and tf record files are are all good now.

One more question. For the pre-training of the mobilenet,  the teacher checkpoint (`TEACHER_CHECKPOINT`) refers to the pre-trained BERT model. Right ? 
If so, then I suppose the pre-trained model should have the configuration in `uncased_L-24_H-1024_B-512_A-4.json` file. But none of the provided pre-trained models on BERT [repo](https://github.com/google-research/bert) has that configuration. Do you know, by any chance, if there is any other place where I can find a pre-trained BERT model with that configuration ?

**Comment by zaidalyafeai** [comment]
@UmarSpa This is the configuration for IB-BERT teacher model in the paper. Unfortunately, the authors didn't publish these models.

**Comment by UmarSpa** [comment]
@zaidalyafeai I see. So essentially I need to train IB-BERT by myself first. And then use this pre-trained model to train MobileBERT. In terms of IB-BERT architecture, it seems the same as MobileBert, just with different configuration.

**Comment by zaidalyafeai** [comment]
@UmarSpa yes, exactly.

**Comment by UmarSpa** [comment]
@zaidalyafeai For the pre-training of IB-BERT, did you use the run_pretraining.py code from MobileBert? or the original run_pretraining.py from Bert?

**Comment by zaidalyafeai** [comment]
You can use either without distillation parameters but note the authors used larger batch sizes with lamb optimizer and higher learning rate.

**Comment by UmarSpa** [comment]
I see. From what I recall, Bert pretraining code did not allow the training on multiple gpus. I wonder if MobileBert does the same. I don't have tpus, so can only use gpus.

**Comment by zaidalyafeai** [comment]
I don't think it allows multiple GPUs because it is just a modified version of the original code of BERT.
